homeassistant/components/vesync/switch.py

```diff
@@ -1,102 +1,105 @@
+            dev_list.append(VeSyncSwitchHA(dev))
+        elif DEV_TYPE_TO_HA.get(dev.device_type) == 'switch':
+            dev_list.append(VeSyncLightSwitch(dev))
         else:
-            count_string = 'switches'
-
-        _LOGGER.info("Discovered %d VeSync %s",
-                     len(manager.devices), count_string)
-
-        for switch in manager.devices:
-            switches.append(VeSyncSwitchHA(switch))
-            _LOGGER.info("Added a VeSync switch named '%s'",
-                         switch.device_name)
-    else:
-        _LOGGER.info("No VeSync devices found")
+            _LOGGER.warning("%s - Unkown device type - %s",
+                            dev.device_name, dev.device_type)
+            continue
 
-    add_entities(switches)
+    async_add_entities(
+        dev_list,
+        update_before_add=True
+    )
 
 
-class VeSyncSwitchHA(SwitchDevice):
+class VeSyncSwitchHA(VeSyncDevice, SwitchDevice):
     """Representation of a VeSync switch."""
 
     def __init__(self, plug):
         """Initialize the VeSync switch device."""
+        super().__init__(plug)
         self.smartplug = plug
-        self._current_power_w = None
-        self._today_energy_kwh = None
 
     @property
-    def unique_id(self):
-        """Return the ID of this switch."""
-        return self.smartplug.cid
-
-    @property
-    def name(self):
-        """Return the name of the switch."""
-        return self.smartplug.device_name
+    def device_state_attributes(self):
+        """Return the state attributes of the device."""
+        attr = {}
+        attr['active_time'] = self.smartplug.active_time
+        if hasattr(self.smartplug, 'weekly_energy_total'):
+            attr['voltage'] = self.smartplug.voltage
+            attr['weekly_energy_total'] = self.smartplug.weekly_energy_total
+            attr['monthly_energy_total'] = self.smartplug.monthly_energy_total
+            attr['yearly_energy_total'] = self.smartplug.yearly_energy_total
+        return attr
 
     @property
     def current_power_w(self):
         """Return the current power usage in W."""
-        return self._current_power_w
+        if hasattr(self.smartplug, 'power'):
+            return self.smartplug.power
+        return None
 
     @property
     def today_energy_kwh(self):
         """Return the today total energy usage in kWh."""
-        return self._today_energy_kwh
+        if hasattr(self.smartplug, 'energy_today'):
+            return self.smartplug.energy_today
+        return None
 
-    @property
-    def available(self) -> bool:
-        """Return True if switch is available."""
-        return self.smartplug.connection_status == "online"
-
-    @property
-    def is_on(self):
-        """Return True if switch is on."""
-        return self.smartplug.device_status == "on"
+    def update(self):
+        """Update outlet details and energy usage."""
+        self.smartplug.update()
+        if hasattr(self.smartplug, 'update_energy'):
```

Comment:
Can we remove this check now and implement the update method in VeSyncLightSwitch too without updating energy?